### PR TITLE
Add option to stop sending read receipts (delabs MSC2285: private read receipts)

### PR DESCRIPTION
**Pull request by SimonBrandner**
Fixes https://github.com/vector-im/element-web/issues/2527
Requires https://github.com/matrix-org/matrix-spec-proposals/pull/2285
Requires https://github.com/matrix-org/matrix-js-sdk/pull/2524
Requires https://github.com/vector-im/element-web/pull/22242
Requires https://github.com/matrix-org/synapse/pull/13273

Notes: Add option to stop sending read receipts (`Preferences` -> `Presence` -> `Send read receipts`)

<hr>

This PR moves `Send read receipts` setting out of labs into `Preferences` -> `Presence`. It also moves `Send typing notification` into that section for consistency.

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

File: src/components/views/elements/SettingsFlag.tsx
Comment (turt2live) [inline src/components/views/elements/SettingsFlag.tsx:36]: This sounds like it should be a boolean, but it's a string - please add docs.

File: src/components/views/elements/SettingsFlag.tsx
Comment (turt2live) [inline src/components/views/elements/SettingsFlag.tsx:107]: this is the line that makes it feel like `disabledTooltip` should be a boolean, so let's avoid the magic syntax in this case:
```suggestion
                        tooltip={this.props.disabled ? this.props.disabledTooltip : null}
```

File: src/components/views/settings/tabs/user/SecurityUserSettingsTab.tsx
Comment (turt2live) [inline src/components/views/settings/tabs/user/SecurityUserSettingsTab.tsx:381]: This heading is confusing - even to users "presence" means online/offline status and not necessarily typing notifications and such.

File: src/components/views/elements/SettingsFlag.tsx
Comment (turt2live) [reply src/components/views/elements/SettingsFlag.tsx:36]: alternatively, rename to `disabledTooltipText` or similar to help self-document

File: src/components/views/settings/tabs/user/SecurityUserSettingsTab.tsx
Comment (turt2live) [inline src/components/views/settings/tabs/user/SecurityUserSettingsTab.tsx:385]: Your*

File: src/settings/Settings.tsx
Comment (turt2live) [inline src/settings/Settings.tsx:392]: Before we transition this feature out of labs, I'd like to see the MSC enter a reasonably positive FCP cycle. Doesn't have to be written into spec, but would prefer to only have to write the feature once without maintaining backwards compatibility.

File: src/components/structures/TimelinePanel.tsx
Comment (turt2live) [inline src/components/structures/TimelinePanel.tsx:938]: I'm not seeing any tests updated for the flag name change: we'll need tests in order to de-labs this.

**Review by turt2live — CHANGES_REQUESTED** [review]

File: src/components/views/settings/tabs/user/SecurityUserSettingsTab.tsx
Comment (SimonBrandner) [reply src/components/views/settings/tabs/user/SecurityUserSettingsTab.tsx:381]: Hmm, I've always interpreted the word as  something that indicates the user's _presence_ on the other side, both RRs and typing notifs would fall under that

That said, I might be just weird in the way I interpreted that - do you have a suggestion for a better heading or should I wait for what design suggests?

File: src/settings/Settings.tsx
Comment (SimonBrandner) [reply src/settings/Settings.tsx:392]: I definitely agree with that; I've made the PR this early to gather feedback from design early to avoid delays on that

File: src/components/views/settings/tabs/user/SecurityUserSettingsTab.tsx
Comment (turt2live) [reply src/components/views/settings/tabs/user/SecurityUserSettingsTab.tsx:381]: other words would be things like "metadata", but might be overly alarmist for the app. Design is definitely a lot better at these sorts of things :)

**Comment by nadonomy** [comment]
@janogarcia I think you're best positioned to take a look at this to carry context forward to the upcoming settings redesign. For now, I think we just want to find a sensible home in our current settings.

From the PR screenshot it looks like it's currently placed in a 'Privacy' section, but I'm not sure this is a natural place to reach. I think it should be placed in the 'Preferences' section.

I've also continually observed users at a loss to what 'Presence' means so think we need to rephrase to more intuitive and predictable language.

File: src/components/structures/TimelinePanel.tsx
Comment (turt2live) [inline src/components/structures/TimelinePanel.tsx:1554]: hmm, I didn't notice this condition in the last review - it's confusing style-wise. Can we make it look like the [style guide](https://github.com/vector-im/element-meta/pull/216)'s approach?

```
if (
  condition
  || condition
) {
  return true;
}
```

I'm less opinionated about the `||` being on the line before or after - would like it broken out either by lines or into variables though.

**Review by turt2live — COMMENTED** [review]
otherwise lgtm code-wise

File: src/components/structures/TimelinePanel.tsx
Comment (SimonBrandner) [reply src/components/structures/TimelinePanel.tsx:1554]: I didn't know we mandated this, so I tend to leave existing code the way it was before but I definitely agree - this is much better!

**Review by turt2live — APPROVED** [review]

**Comment by janogarcia** [comment]
@nadonomy @SimonBrandner 

On the one hand, I'd say that Privacy could be a natural place for it. But, on the other hand, it would still be a bit buried due to the current settings layout; that's why we may want to promote it to the Preferences tab instead.

As for the section name, I think `Presence` could actually work. Adding a contextual note (`Choose what real-time activity other users can see from you.  `) would help make it more obvious. Alternatively, we could change the label to `Visibility`.

What about adding something along these lines above (or below) the `Preferences > Composer` section? 

---

**Presence **
Choose what real-time activity other users can see from you.  

Send typing notifications     ✅
Send read receipts              ✅

---

**Visibility **
Choose what real-time activity other users can see from you.  

Send typing notifications     ✅
Send read receipts              ✅

**Review by janogarcia — COMMENTED** [review]
Refer to previous comment. https://github.com/matrix-org/matrix-react-sdk/pull/8629#issuecomment-1154086499

**Review by turt2live — CHANGES_REQUESTED** [review]
I didn't connect that there's a de-labs PR and that the MSC wasn't in FCP. For code review, we'll want to make sure https://github.com/matrix-org/matrix-spec-proposals/pull/2285 has at least entered FCP before merging.

Marking as requesting changes until the MSC makes movement. If it takes too long, let's talk about options oob.

**Comment by SimonBrandner** [comment]
@janogarcia, thank you for the review! I've updated the PR:

**Review by janogarcia — CHANGES_REQUESTED** [review]
Let's change the copy to `Choose what others can see from your presence.`.

**Unsupported disabling read receipts**

Also, forgot to comment that disabled states shouldn't be interactive, thus, please remove the tooltip and place that copy (`Your server doesn't support disabling sending read receipts.`) in a `. mx_SettingsFlag_microcopy` secondary line instead, as in the toggle switch for Spaces (`All rooms you're in will appear in Home.`).

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (turt2live) [inline src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:57]:
```suggestion
    public static PRESENCE_SETTINGS = [
```

Comment (turt2live) [reply src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:57]: please fix the others while we're here

**Review by turt2live — CHANGES_REQUESTED** [review]
code lgtm, but abusing review status to block from merging (needs FCP to have more checkboxes)

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (SimonBrandner) [reply src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:57]: It feels like these should be private which is weird... So I think perhaps simple `const`s should work fine?

**Comment by SimonBrandner** [comment]
> code lgtm, but abusing review status to block from merging (needs FCP to have more checkboxes)

I am aware - that's why there's the `X-Blocked`

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (turt2live) [reply src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:57]: as members of the class I don't think they can be `const`, but `private` would be better than what they currently are :)

Comment (SimonBrandner) [reply src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:57]: I meant moving them out of the class

Comment (turt2live) [reply src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:57]: ah, let's not do that in this PR so the variables are consistent. Moving them all out of all the related classes would be a whole other PR.

**Review by turt2live — APPROVED** [review]
lgtm pending product

**Comment by SimonBrandner** [comment]
Internal context: due to several things we have decided not to release this out of labs just yet and just switch to the stable prefixes 

For delabs PR, see https://github.com/matrix-org/matrix-react-sdk/pull/9128

File: src/components/structures/TimelinePanel.tsx
Comment (turt2live) [inline src/components/structures/TimelinePanel.tsx:992]: since we're in an async function, we'd normally change this to a `try/catch`, but for this PR I don't think it's important.

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (turt2live) [inline src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:61]:
```suggestion
        // sendReadReceipts - handled specially due to server needing support
```

**Review by turt2live — APPROVED** [review]
overall looks good to me. 

Product/QA: if a testing environment is needed, please ping me.

File: src/components/structures/TimelinePanel.tsx
Comment (SimonBrandner) [reply src/components/structures/TimelinePanel.tsx:992]: (I mean are there even server that don't support `/read_markers` at this point?)

Comment (turt2live) [reply src/components/structures/TimelinePanel.tsx:992]: probably, but also I would argue we don't support them.

File: test/test-utils/test-utils.ts
Comment (turt2live) [inline test/test-utils/test-utils.ts:129]: linter seems to be complaining about this because it's defined 2 lines lower

Comment (SimonBrandner) [reply test/test-utils/test-utils.ts:129]:
```suggestion
```

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (kittykat) [inline src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:211]: What about "Share your activity and status with others."?

**Review by kittykat — CHANGES_REQUESTED** [review]

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (SimonBrandner) [reply src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:211]: Sounds great!!

**Review by kittykat — APPROVED** [review]

**Comment by cdauth** [comment]
I am unfamiliar with the release process of Element and how it relates to matrix-react-sdk. Could someone explain whether this is or will be part of both Element Web and Element Android, and if so whether it has already been released or when it will be released?

**Comment by t3chguy** [comment]
Element Android is a separate codebase. This has been in Element Web & Desktop for multiple versions.

image

Your server must advertise support for it though.

**Pull request merged**
